### PR TITLE
Make Phobos compilable with preview=fieldwise

### PR DESCRIPTION
**Pull request by nordlow**
Reopened to see if things build.

### CODE REVIEW

**Comment by nordlow** [comment]
Alternatively, is it better to statically check whether an `==`-expression constant folds or not?

**Comment by nordlow** [comment]
No correctly handles opEquals aswell, @MoonlightSentinel.

**Review by thewilsonator — APPROVED** [review]

**Comment by MoonlightSentinel** [comment]
Nice, that seems to work with and without the flag. But I'm not sure if we should enable it by default for all CI machines given that code could break when not compiling with `-preview=fieldwise`.

Let's merge the bugfix first and deal with the transition in the dmd pr when it becomes the default.

**Comment by nordlow** [comment]
Do you mean that this PR is the bugfix or you should I remove anything from this PR? I'm satisfied.

**Comment by MoonlightSentinel** [comment]
I meant the changes to the Makefiles, we want to ensure Phobos is compilable without `-preview=fieldwise`.

File: std/variant.d [hunk 1]
```diff
@@ -249,11 +249,21 @@ private:
...
             {
-                if (*rhsPA == *zis)
+                enum isEmptyStructWithoutOpEquals = (is(A == struct) && A.tupleof.length == 0 &&
+                                                     !__traits(hasMember, A, "opEquals"));
+                static if (!isEmptyStructWithoutOpEquals)
+                {
+                    if (*rhsPA == *zis)
+                        return 0;
+                }
+                static if (isEmptyStructWithoutOpEquals)
```
Comment (andralex) [inline std/variant.d:259]:
```suggestion
                else
```

File: std/variant.d [hunk 2]
```diff
@@ -249,11 +249,21 @@ private:
...
                 {
                     return 0;
+                    // The check above will always succeed if A is an empty struct.
```
Comment (andralex) [inline std/variant.d:262]: Would the comment usually go before the code?

File: std/variant.d [hunk 3]
```diff
@@ -249,11 +249,21 @@ private:
         {
             static if (is(typeof(*rhsPA == *zis)))
             {
-                if (*rhsPA == *zis)
+                enum isEmptyStructWithoutOpEquals = (is(A == struct) && A.tupleof.length == 0 &&
+                                                     !__traits(hasMember, A, "opEquals"));
```
Comment (andralex) [inline std/variant.d:253]:
```suggestion
                enum isEmptyStructWithoutOpEquals = is(A == struct) && A.tupleof.length == 0 &&
                                                     !__traits(hasMember, A, "opEquals");
```

**Comment by nordlow** [comment]
> Nice, that seems to work with and without the flag. But I'm not sure if we should enable it by default for all CI machines given that code could break when not compiling with `-preview=fieldwise`.

Shall we revert the additions of `-preview=dtorfields` aswell?

File: std/variant.d (hunk 2 above)
Comment (nordlow) [reply std/variant.d:262]: Fixed.

**Comment by wilzbach** [comment]
> Shall we revert the additions of `-preview=dtorfields` aswell?

No that's fine and very unlikely to break something. Furthermore, that one is more of a bug fix.

File: std/variant.d (hunk 1 above)
Comment (nordlow) [reply std/variant.d:259]: This doen't work because there's an `else static if` after this `static if`.

File: std/variant.d (hunk 3 above)
Comment (nordlow) [reply std/variant.d:253]: Done.

**Comment by nordlow** [comment]
Corrected according to comments.

File: std/variant.d (hunk 1 above)
Comment (andralex) [reply std/variant.d:259]: Not a big deal, but just seeing a condition followed by its negation seemed odd. The refactoring that would take care of that would be:

```d
                static if (isEmptyStructWithoutOpEquals)
                {
                    // The check above will always succeed if A is an empty struct.
                    // Don't generate unreachable code as seen in
                    // https://issues.dlang.org/show_bug.cgi?id=21231
                    return 0;
                }
                else
                {
                    if (*rhsPA == *zis)
                        return 0;
                    static if (is(typeof(*zis < *rhsPA)))
                    {
                        // Many types (such as any using the default Object opCmp)
                        // will throw on an invalid opCmp, so do it only
                        // if the caller requests it.
                        if (selector == OpID.compare)
                            return *zis < *rhsPA ? -1 : 1;
                        else
                            return ptrdiff_t.min;
                    }
                    else
                    {
                        // Not equal, and type does not support ordering
                        // comparisons.
                        return ptrdiff_t.min;
                    }
            }
```

This better clarifies the flow depending on compile-time conditions.

Comment (nordlow) [reply std/variant.d:254]: done.

**Comment by nordlow** [comment]
Ping. Ready to roll?

**Comment by nordlow** [comment]
AFAICT, [opPostMove](https://github.com/dlang/DIPs/pull/109) will be possible if we `-preview=fieldwise` becomes the default.

**Comment by nordlow** [comment]
Reverted changes to `win32.mak` and `win64.mak`. Ready for merge?

**Pull request merged**